### PR TITLE
Update internal timers to use NewTimerWithOptions

### LINKED ISSUE #1602: Update internal timers to use NewTimerWithOptions and a summary
**Describe the solution you'd like**

Since #1597, there is now a way to provide a summary for a timer, so we should do so for internal timers we create inside the SDK

### PR DESCRIPTION
**Pull request by yuandrew**
## What was changed

- Change internal uses of `NewTimer` to `NewTimerWithOptions`
- Created new `AwaitWithTimeoutAndOption` to allow specifying options for the underlying `timer` that's created

## Why?

Because there's now a way to give a summary for internal timers, change internal code to use the new timer with summary command.

## Checklist

1. Closes #1602 

2. How was this tested:

3. Any docs updates needed?

### CODE REVIEW

File: workflow/deterministic_wrappers.go
Comment (yuandrew) [inline workflow/deterministic_wrappers.go:108]: Does changing this function signature break anything/anyone external?

File: internal/interceptor.go
Comment (Quinn-With-Two-Ns) [inline internal/interceptor.go:214]: We cannot change the API like this unfortunately

Comment (Quinn-With-Two-Ns) [reply internal/interceptor.go:214]: We will need to add a new API to pass options

File: workflow/deterministic_wrappers.go
Comment (Quinn-With-Two-Ns) [reply workflow/deterministic_wrappers.go:108]: Yes it would break all callers when they update their SDK. That is not okay. Go lang doesn't give us a lot of tools to work with here so we need to just add a new API like `AwaitWithOptions` or `AwaitWithTimeoutAndOptions`

Comment (yuandrew) [reply workflow/deterministic_wrappers.go:108]: had a hunch that'd be the case

File: internal/workflow.go
Comment (cretz) [inline internal/workflow.go:383]: I don't think we need to change anything related to static summary/details or child workflows in this timer-specific issue. Also, while that issue is technically fixed, it would need to be in the dev server and we need integration tests.

File: internal/workflow.go
Comment (cretz) [inline internal/workflow.go:508]:
```suggestion
	timer := NewTimerWithOptions(ctx, timeout, TimerOptions{Summary: "AwaitWithTimeout"})
```

This is clearer to readers and a bit safer

File: internal/workflow.go
Comment (yuandrew) [reply internal/workflow.go:383]: sounds good, I'll take these changes out of this PR

File: internal/workflow.go
Comment (yuandrew) [inline internal/workflow.go:527]: not sure why someone might call `AwaitWithTimeoutAndOptions` without specifying a summary, but defaulting to "AwaitWithTimeoutAndOptions"

Comment (cretz) [reply internal/workflow.go:527]: I'm 50/50 here. We do not default a summary in `NewTimerWithOptions`, not sure we need to here. But then again it makes some sense since we are in the option-less equivalent calls. No strong opinion here.

Comment (yuandrew) [reply internal/workflow.go]: After some thought, i think I'm leaning the other way now, if someone chooses to use `AwaitWithTimeoutAndOptions`, they should probably be responsible for setting their `TimerOptions` properly

File: internal/interceptor.go
Comment (Quinn-With-Two-Ns) [inline internal/interceptor.go:219]: I wonder if we should just call this `AwaitWithOptions` and take timeout as an optional option?

File: internal/workflow.go
Comment (Quinn-With-Two-Ns) [reply internal/workflow.go]: Yeah I'd be consistent with `NewTimerWithOptions` . We should set some sensible default if the user is calling `AwaitWithTimeout`

File: internal/interceptor.go
Comment (yuandrew) [reply internal/interceptor.go:219]: As in move `timeout` into the `TimerOptions` struct? If that's what you mean, do we want to change the `NewTimerWithOptions` to match?

Comment (Quinn-With-Two-Ns) [reply internal/interceptor.go:219]: More like create `AwaitOptions` and put `timeout` and `TimerOptions` under it. This is so if we need to add more fields we don't have to add more overrides

File: workflow/deterministic_wrappers.go
Comment (cretz) [inline workflow/deterministic_wrappers.go:125]: You'll need a type alias for `AwaitOptions` in the public package here to reference the internal, otherwise users can't even call it. This may surface if/when you write integration tests.

Comment (yuandrew) [reply workflow/deterministic_wrappers.go:125]: Is it worth adding a unit test for `AwaitWithOptions` that does the same thing as `AwaitWithTimeout` just to exercise the new function codepath in our tests?

Comment (cretz) [reply workflow/deterministic_wrappers.go]: Doesn't necessarily do the same thing, it also allows you to add a summary. Might as well test it, but I don't have a strong opinion. I think it'd have caught that certain needed things weren't exposed here though.

File: test/integration_test.go
Comment (cretz) [inline test/integration_test.go:6275]: Can we remove this print?

**Review by cretz — APPROVED** [review]
LGTM, may want to wait until @Quinn-With-Two-Ns's approval though

**Review by Quinn-With-Two-Ns — APPROVED** [review]

**Comment by cretz** [comment]
@yuandrew - Looks like you can merge if/when CI passes (I kicked it again, looks like flaky test)

**Comment by yuandrew** [comment]
@cretz  looks like there's a regression with the `TestWorkerVersioningTestSuite/TestCommitRules` test, looking to repro this locally

**Pull request merged**
